### ISSUE TITLE
Updating key is more expensive than inserting a new one

### ISSUE THREAD

**Issue by nekto0n** [coreos/etcd#6615]
If we modify `BenchmarkWatchableStoreTxnPut` to use values that look more real
```diff
--- a/mvcc/watchable_store_bench_test.go
+++ b/mvcc/watchable_store_bench_test.go
@@ -50,9 +50,8 @@ func BenchmarkWatchableStoreTxnPut(b *testing.B) {
        defer cleanup(s, be, tmpPath)
 
        // arbitrary number of bytes
-       bytesN := 64
-       keys := createBytesSlice(bytesN, b.N)
-       vals := createBytesSlice(bytesN, b.N)
+       keys := createBytesSlice(64, b.N)
+       vals := createBytesSlice(2048, b.N)
 
        b.ResetTimer()
        b.ReportAllocs()
```
we get something like this
```
BenchmarkWatchableStoreTxnPut-4        30000         50327 ns/op        7729 B/op         14 allocs/op
```
If we modify this benchmark to use *same key* every time, we'll get these numbers:
```
BenchmarkWatchableStoreTxnPut-4        30000         60183 ns/op       10062 B/op         20 allocs/op
```
Which is about 20% more expensive in allocations.

Extra allocations come [this branch in put method](https://github.com/coreos/etcd/blob/master/mvcc/kvstore.go#L521), which extracts `KeyValue` from boltdb and deserializes whole struct just to get lease id. I think that:
  * `kv.Lease` is almost always `0`
  * extracting whole object (which in my case about 2Kb) just to get 8 bytes value is an overkill.

I have some thoughts how this can be fixed:
- store lease in `kvindex` along revision
- serialize lease in boltdb separately from `KeyValue`. We can prepend/append it to bytes representation, but this will require changing on-disk format.
- write partial deserialization routine to extract only `Lease` field from serialized protobuf. It can be fragile and still will require calls to boltdb

What do you think?

**Comment by xiang90** [comment]
> store lease in kvindex along revision

I think this is probably the best way to go.

**Comment by xiang90** [comment]
@nekto0n BTW, I am actually not super worry about the put allocation. All etcd put requests are serialized, so we can do 1 put a time. We expect the put throughput to be 10,000s/second. You can try benchmark tool to know the exact improvement after improving the allocation. 

I care more about the watch allocation, since we have have N watchers on one key. A put can trigger N watch response. And we expect users to put more than one watcher on a key.

**Comment by nekto0n** [comment]
In my case we have like 99.999% of updates and only a few watchers. So I'd like to optimize this case a bit. We may be planning to use a lot of watchers in future, optimizing that path would be great too :)

**Comment by nekto0n** [comment]
About implementation details. I see two choices:
- Add `leaseID` field to `type revision struct`
- Add new `type revInfo struct`:
```
type revInfo struct {
	revision
	leaseID lease.LeaseID
}
```
Former seems to make `revision` not a good name for this struct, while latter brings some burden to initialization everywhere especially in tests, e.g.
```
allRevs := []revision{{4, 0}, {6, 0}, {8, 0}, {10, 0}, {12, 0}, {14, 1}, {16, 0}}
```
becomes
```
allRevs := []revInfo{
		{revision{4, 0}, 0},
		{revision{6, 0}, 0},
		{revision{8, 0}, 0},
		{revision{10, 0}, 0},
		{revision{12, 0}, 0},
		{revision{14, 1}, 0},
		{revision{16, 0}, 0},
	}
```
Any thoughts?

**Comment by xiang90** [comment]
@nekto0n I think we only need to remember the latest leaseID. So adding a leaseID into keyIndex sttruct would be good enough?

**Comment by heyitsanthony** [comment]
Another option: keep a key->id map in lessor and change the lessor interface to have `Attach(LeaseItem)` and `Detach(LeaseItem)`? Seems like a more convincing design than reworking kvindex to have non-revision data.

**Comment by nekto0n** [comment]
@heyitsanthony I think that's a good approach and we won't have to pay extra 8 bytes per every `keyIndex`, only allocating proportional to the number of leased keys. I'll try that.

**Comment by nekto0n** [comment]
I guess this one is fixed in #6620
